platform/openide.util.ui/src/org/openide/util/ImageUtilities.java

```
@@ -185,31 +199,87 @@ public static final Image loadImage(String resourceID) {
+     *         available
+     * @since 7.36
+     */
+    public static final Image loadImage(URI uri) {
+        Parameters.notNull("icon", uri);
+        String scheme = uri.getScheme();
+        if (scheme.equals("nbresloc")) { // NOI18N
+            // Apply our dedicated handling logic. Omit the initial slash of the path.
+            return loadImage(uri.getPath().substring(1), false);
+        } else {
+            if (!(scheme.equals("file") ||
+                scheme.equals("jar") && uri.toString().startsWith("jar:file:") ||
+                scheme.equals("file")))
+            {
+                LOGGER.log(Level.WARNING, "loadImage(URI) called with unusual URI: {0}", uri);
+            }
+            try {
+                /* Observed to return an image with size (-1, -1) if URL points to a non-existent
+                file (after ensureLoaded(Image) is called). */
+                return Toolkit.getDefaultToolkit().createImage(uri.toURL());
+            } catch (MalformedURLException e) {
+                LOGGER.log(Level.WARNING, "Malformed URL passed to loadImage(URI)", e);
+                return null;
+            }
+        }
+    }
+
+    /**
+     * Convert an {@link Icon} instance to a delegating {@link ImageIcon} instance. If the supplied
+     * Icon instance is an SVG icon or has other HiDPI capabilities provided by the methods in this
+     * class, they are preserved by the conversion.
+     *
+     * <p>This method is intended for use only when existing APIs require the use of ImageIcon. In
+     * most other situations it is preferable to use the more general Icon type.
+     *
+     * @param icon the icon to be converted
+     * @return the converted instance, or null if {@code icon} was null
+     * @since 7.36
+     */
+    public static final ImageIcon toImageIcon(Icon icon) {
+        if (icon == null) {
+            return null;
+        }
```

it might be better to fail on null since (new) code which doesn't check for null Icons before passing it to a conversion method is likely buggy and might benefit from failing early.

`image2Icon` will fail on null i believe, `icon2Image` does accept null but will log a warning at least.

what do you think about keeping the naming pattern? -> `icon2ImageIcon`?

---

When doing bulk refactoring in the style of https://github.com/apache/netbeans/pull/8109, it is useful if `icon.someMethod()` and `toImageIcon(icon).someMethod()` is guaranteed not to change the current behavior. Otherwise, existing features could break e.g. if they previously relied on null values being quietly accepted.

(Often I can read the existing code and be 95% sure a null value is never passed... but getting the last 5% of certainty is very difficult. And I don't want to have to insert null checks everywhere, if I'm changing 20 call sites and each case is really not supposed to be null in the first place.)

Perhaps I will specify in the Javadoc that null is not a valid input, but log a warning if a null does occur, and still return null in that unexpected case?

> what do you think about keeping the naming pattern? -> icon2ImageIcon?

Good idea, will rename.

---

I'll take the "specify that null is not allowed, but log a warning and return null if it is passed anyway" approach here.